services/brig/test/integration/API/User/Auth.hs

```diff
@@ -456,6 +457,52 @@ testLoginVerify6DigitResendCodeSuccessAndRateLimiting brig galley _opts db = do
         (Just defCookieLabel)
         (Just $ Code.codeValue mostRecentCode)
 
+testLoginVerify6DigitLimitRetries :: Brig -> Galley -> Opts.Opts -> DB.ClientState -> Http ()
+testLoginVerify6DigitLimitRetries brig galley _opts db = do
+  (u, tid) <- createUserWithTeam' brig
+  let Just email = userEmail u
+  let checkLoginFails body =
+        login brig body PersistentCookie !!! do
+          const 403 === statusCode
+          const (Just "code-authentication-failed") === errorLabel
+
+  Util.setTeamFeatureLockStatus @Public.SndFactorPasswordChallengeConfig galley tid Public.LockStatusUnlocked
+  Util.setTeamSndFactorPasswordChallenge galley tid Public.FeatureStatusEnabled
+  Util.generateVerificationCode brig (Public.SendVerificationCode Public.Login email)
+  key <- Code.mkKey (Code.ForEmail email)
+  Just correctCode <- Util.lookupCode db key Code.AccountLogin
+  let wrongCode = Code.Value $ unsafeRange (fromRight undefined (validate "123456"))
+  -- login with wrong code should fail 3 times
+  checkLoginFails $
+    PasswordLogin $
+      PasswordLoginData
+        (LoginByEmail email)
+        defPassword
+        (Just defCookieLabel)
+        (Just wrongCode)
+  checkLoginFails $
+    PasswordLogin $
+      PasswordLoginData
+        (LoginByEmail email)
+        defPassword
+        (Just defCookieLabel)
+        (Just wrongCode)
+  checkLoginFails $
+    PasswordLogin $
+      PasswordLoginData
+        (LoginByEmail email)
+        defPassword
+        (Just defCookieLabel)
+        (Just wrongCode)
+  -- after 3 failed attempts, login with correct code should fail as well
+  checkLoginFails $
```

Comment:
I'm wondering if we could reduce code duplication by extracting a function: `checkLoginFailsWithCode :: MonadHttp m => Code -> m ()` :thinking: